text/0000-module-unification.md

```diff
@@ -0,0 +1,829 @@
+and `routes` collections.
+
+Ember CLI will have a build step that normalizes modules to a common form and
+builds a mapping between that form and the ES module path described above.
+While building this normalized map, the build must error and provide useful
+messages if any module naming errors are detected. Unregistered collections and
+types should not be allowed. Also, the same normalized module path must not be
+repeated through alternative naming forms.
+
+### Module Type
+
+The type of a module can be determined through the following file naming and
+module export rules:
+
+1. `src/${type}` - typed modules named `main` (explained further below), in
+    which default exports match the type specified by the file name.
+2. `src/${collection}/${namespace}/${name}/${type}` - expanded collection
+    modules, in which default exports match the type specified by the file name.
+3. `src/${collection}/${namespace}/${name}` - in which type can be inferred
+    based on the module's exports. Default exports must match the default
+    type for the collection. If there is no default export, named exports will
+    be scanned for a matching type allowed in the collection.
+
+Note that template precompilers will need to use default vs. named exports
+appropriately in order to satisfy the expectations of Rules 2 and 3.
+
+Here are a few example applications of the module type determination rules:
+
+```
+// Rule 1
+
+src/router (with `export default Ember.Router.extend()`)
+=> name: 'main',
+   type: 'router'
+
+// Rule 2
+
+src/ui/routes/posts/post/route.js (with `export default Ember.Route.extend()`)
+=> collection: 'ui/routes',
+   namespace: 'posts',
+   name: 'post',
+   type: 'route'
+
+src/ui/routes/posts/post/template (with `export default Ember.HTMLBars.template(COMPILED)`)
+=> collection: 'ui/routes',
+   namespace: 'posts',
+   name: 'post',
+   type: 'template'
+
+// Rule 3
+
+src/data/models/author (with `export default DS.Model.extend()`)
+=> collection: 'data/models',
+   name: 'author',
+   type: 'model' (the default type for the models collection)
+
+src/ui/components/titleize (with `export let helper = Ember.Helper.helper(function() { })`)
+=> collection: 'ui/components',
+   name: 'titleize',
+   type: 'helper'
+
+src/ui/components/show-title (with `export let template = Ember.HTMLBars.template(COMPILED)`)
+=> collection: 'ui/components',
+   name: 'show-title',
+   type: 'template'
+```
+
+### Main Modules
+
+Every project must have a "main" module, named `src/main.js`, that
+serves as an entry point into the project.
+
+The main module must export an `Application`, `Engine`, or (new) `Addon` class.
+This class must define a `modulePrefix`, which must match the node package name
+for the project.
+
+The main module also declares other properties that help the Ember resolver
+understand relationships between projects. For instance, the main module can
+declare which modules in an addon are available to a consuming app's resolver.
```

Comment:
It will absolutely be possible to create a custom resolver that extends from the resolver changes proposed here. I would assume that your current custom resolver would need to be rewritten to take advantage of the new system, but it should be straightforward.